### PR TITLE
Add backoff procedure to universe RPC proof courier

### PR DESCRIPTION
**Pull request by ffranr**
This PR extracts the backoff procedure from the hashmail proof courier and makes use of it in the universe RPC proof courier.

Fixes https://github.com/lightninglabs/taproot-assets/issues/512

### CODE REVIEW

File: proof/courier.go [hunk 1]
```diff
@@ -513,15 +514,16 @@ type Recipient struct {
...
+// or not.
+type BackoffExecError struct {
+	execErr error
+}
 
-	// BackoffCfg configures the behaviour of the proof delivery
-	// functionality.
-	BackoffCfg *BackoffCfg
+func (e *BackoffExecError) Error() string {
+	return fmt.Sprintf("backoff exec error: %s", e.execErr.Error())
```
Comment (guggero) [inline proof/courier.go:536]: Do we need a `nil` check on the `execErr` here? Or can we be reasonably sure it is always set?

**Comment by ffranr** [comment]
I haven't added an itest which covers this new use of the backoff procedure in the universe proof courier. Let me know if you guys have any ideas along those lines.

File: proof/courier.go [hunk 2]
```diff
@@ -543,21 +545,190 @@ type BackoffCfg struct {
 	MaxBackoff time.Duration `long:"maxbackoff" description:"The maximum backoff time to wait before retrying to deliver the proof to the receiver."`
 }
 
+// BackoffHandle is a handle to the backoff procedure.
+type BackoffHandle struct {
```
Comment (guggero) [inline proof/courier.go:564]: nit: since this is a concrete type/struct, wouldn't it be a handle**r** (instead of just handle)? I initially thought this was a typo but it seems very consistent throughout the diff.

File: proof/courier.go [hunk 3]
```diff
@@ -543,21 +545,190 @@ type BackoffCfg struct {
...
+	timestamps, err := b.deliveryLog.QueryProofDeliveryLog(
+		ctx, proofLocator,
+	)
+	if err != nil {
+		return fmt.Errorf("unable to retrieve proof delivery "+
+			"logs: %w", err)
+	}
+
+	if len(timestamps) == 0 {
+		log.Infof("No previous proof delivery attempts found "+
```
Comment (guggero) [inline proof/courier.go:601]: This sounds like a `Debugf` or even `Tracef` log.

Comment (guggero) [reply proof/courier.go:601]: Same for the other log outputs in this method (info should be important enough for the user to actually make it into the log).

File: proof/courier.go [hunk 4]
```diff
@@ -543,21 +545,190 @@ type BackoffCfg struct {
...
+	BackoffCfg *BackoffCfg
+}
+
 // HashMailCourier is a hashmail proof courier service handle. It implements the
 // Courier interface.
 type HashMailCourier struct {
 	// cfg contains the courier's configuration parameters.
 	cfg *HashMailCourierCfg
 
+	// BackoffHandle handles the backoff procedure for the proof delivery
```
Comment (guggero) [inline proof/courier.go:758]: nit: Godoc comment mismatch (case).

File: proof/courier.go [hunk 5]
```diff
@@ -581,89 +752,50 @@ func (h *HashMailCourier) DeliverProof(ctx context.Context,
...
-	//
-	// Only wait if we have a non-zero number of past delivery attempts.
-	timeSinceLastAttempt := timeSinceLastDeliveryAttempt(timestamps)
-	backoffResetWait := h.cfg.BackoffCfg.BackoffResetWait
-	if len(timestamps) > 0 &&
-		timeSinceLastAttempt < backoffResetWait {
+		// Now that the stream has been initialized, we'll write
+		// the proof over the stream.
+		//
+		// TODO(roasbeef): do ecies here
```
Comment (guggero) [inline proof/courier.go:804]: What could the TODO here mean? Maybe a typo of "retries"? Then I guess it would be addressed now.

File: proof/courier.go [hunk 6]
```diff
@@ -622,10 +639,17 @@ func (b *BackoffHandle) Exec(ctx context.Context, proofLocator Locator,
 		return fmt.Errorf("backoff config not specified")
 	}
 
+	locatorHash, err := proofLocator.Hash()
+	if err != nil {
+		return err
+	}
+	log.Infof("Starting proof delivery backoff procedure for proof: "+
+		"(locator_hash=%x)", locatorHash)
```
Comment (guggero) [inline proof/courier.go:647]: I think this logs as double hashed, since it's a hash that has a `String()` method. So should proably be `locatorHash[:]` instead.

File: proof/courier.go [hunk 7]
```diff
@@ -984,6 +1008,10 @@ type UniverseRpcCourier struct {
 	// the universe RPC server.
 	client unirpc.UniverseClient
 
+	// BackoffHandle handles the backoff procedure for the proof delivery
```
Comment (guggero) [inline proof/courier.go:1011]: nit: same here re godoc comment case mismatch.

**Review by guggero — APPROVED** [review]
Very nice! Looks great, nothing much to add other than nits.

I think as a follow-up PR we should enable the RPC universe courier by default in the itests and remove all calls to `sendProof()` (which circumvents any courier at the moment).

File: proof/courier.go (hunk 1 above)
Comment (jharveyb) [reply proof/courier.go:536]: AFAICT `BackoffHandle.Exec()` is the only place we're building such an error. But a nil check would be good anyways, could just return "backoff exec error" imo.

File: proof/courier.go (hunk 3 above)
Comment (jharveyb) [reply proof/courier.go:601]: Agree, I think with that change details about the backoff would only get logged on failure, with the final error about # of attempts.

File: proof/courier.go (hunk 5 above)
Comment (jharveyb) [reply proof/courier.go:804]: Not addressed, related to encrypting proofs for the receiver before uploading to the courier:

https://github.com/lightninglabs/taproot-assets/issues/478

**Review by jharveyb — COMMENTED** [review]
Looks good, nice refactor to make this reusable!

I think one way to add this into the itests would just be to have a test that's a duplicate of `testReattemptFailedAssetSend`, but with `proofCourierType: proof.UniverseRpcCourierType`.

When I do that, the test hangs as the transferEvents are never broadcast. Issue may be in `deliverFunc()` but I didn't dig in further.

File: proof/courier.go (hunk 5 above)
Comment (ffranr) [reply proof/courier.go:804]: I've added an additional comment to this TODO

**Comment by ffranr** [comment]
The latest commits address the review comments and add an itest to cover the changes introduced in this PR.

**Review by jharveyb — APPROVED** [review]
Good to see the itest changes were small! LGTM 👍🏽

File: proof/courier.go [hunk 8]
```diff
@@ -1012,20 +1106,27 @@ func (c *UniverseRpcCourier) DeliverProof(ctx context.Context,
...
 
-		// Submit proof to courier.
-		_, err = c.client.InsertProof(ctx, &unirpc.AssetProof{
-			Key:       &universeKey,
-			AssetLeaf: &assetLeaf,
-		})
+		// Setup delivery routine and start backoff procedure.
+		deliverFunc := func() error {
+			// Submit proof to courier.
+			_, err = c.client.InsertProof(ctx, &unirpc.AssetProof{
```
Comment (Roasbeef) [inline proof/courier.go:1113]: This only adds back off for proof insertion. What we've seen fail is that the receiver attempts to look up, but maybe got the block before the sender, then never retries again.

We also need the receiver to properly retry.

File: itest/send_test.go [hunk 9]
```diff
@@ -492,6 +501,109 @@ func testReattemptFailedAssetSend(t *harnessTest) {
...
+	AssertAddrCreated(t.t, t.tapd, rpcAssets[0], recvAddr)
+
+	// Simulate a failed attempt at sending the asset proof by stopping
+	// the proof courier service.
+	//
+	// In following the hashmail proof courier protocol, the receiver node
+	// returns a proof received confirmation message via the courier.
+	// We can simulate a proof transfer failure by stopping the receiving
+	// tapd node. The courier service should still be operational.
+	require.NoError(t.t, t.proofCourier.Stop())
```
Comment (Roasbeef) [inline itest/send_test.go:598]: How is this testing the back off? We want to make sure that if the receiver gets there too early, they continue to retry in the background.

**Review by Roasbeef — CHANGES_REQUESTED** [review]

File: itest/send_test.go (hunk 9 above)
Comment (jharveyb) [reply itest/send_test.go:598]: I think this does properly test it for the sender as-is.

Comment (guggero) [reply itest/send_test.go:598]: I was a bit confused at the beginning as well. But basically what this test does is test that we get a `ReceiverProofBackoffWaitEvent` event if the proof courier isn't started. Meaning the sender tries to send, fails then backs off and tries _again_.
But the test doesn't currently assert the successful submission.

IMO we can add the backoff-retry mechanism for the receiver in a separate PR and there extend both of these two tests to make sure that the transmission eventually works if the proof courier is turned on.

**Review by guggero — APPROVED** [review]
Nice, LGTM :tada: 

I suggest doing the backoff-retry for the _receiver_ in a separate PR and get this in as is.
